[r2] add troubleshooting (#17623)
cloudflare/cloudflare-docs: src/content/docs/r2/platform/troubleshooting.mdx

---
pcx_content_type: troubleshooting
title: Troubleshooting

---

## Troubleshooting 403 / CORS issues with R2

If you are encountering a CORS error despite setting up everything correctly, you may follow this troubleshooting guide to help you.

> In your browser console, do you see a 401/403 error right above the CORS error? If yes, then you aren't actually dealing with a CORS issue! If you do have a CORS issue, head to the section on [resolving CORS issues](#if-its-actually-cors).

Reviewer: ```suggestion
If you see a 401/403 error above the CORS error in your browser console, you are dealing with a different issue (not CORS related).

If you do have a CORS issue, refer to [Resolving CORS issues](#if-its-actually-cors).
```

You need to tweak your Hotlink Protection settings using a [Configuration Rule](/rules/configuration-rules/) or you can disable it completely.

### If you are using the S3 API

Your request is very likely incorrectly signed.
You may see a better error message by trying the request over curl.

You can find working S3 signing examples on the [Examples](/r2/examples/aws/) page.

### If it's actually CORS

Here are some common issues with CORS configurations:

- `ExposeHeaders` is missing headers like `ETag`
- `AllowedHeaders` is missing headers like `Authorization` or `Content-Type`
- `AllowedMethods` is missing methods like `POST`/`PUT` (you don't need to include `OPTIONS` fyi)

## HTTP 5XX Errors and Capacity limitations of Cloudflare R2

When you encounter an HTTP 5XX error, it's usually a sign that your Cloudflare R2 bucket has been overwhelmed by too many concurrent requests.
These errors can trigger bucket-wide read and write locks, affecting the performance of all ongoing operations.
To avoid these disruptions, it's important to implement strategies for managing request volume.

Here are some mitigations you can employ:

### Monitor Concurrent Requests

> Track the number of concurrent requests to your bucket. If a worker encounters a 5XX error, ensure that it retries the operation and communicates with other workers.  By coordinating, workers can collectively slow down, reducing the request rate and maintaining a more stable flow of successful operations.

Author: workers here isn't referring to Cloudflare Workers

---
pcx_content_type: troubleshooting
title: Troubleshooting

---

## Troubleshooting 403 / CORS issues with R2

If you are encountering a CORS error despite setting up everything correctly, you may follow this troubleshooting guide to help you.

If you see a 401/403 error above the CORS error in your browser console, you are dealing with a different issue (not CORS related).

> If you do have a CORS issue, refer to [Resolving CORS issues](#if-its-actually-cors).

Reviewer: ```suggestion
If you do have a CORS issue, refer to [Resolving CORS issues](#if-it-is-actually-cors).
```

> ### If you are using a Custom domain

Reviewer: ```suggestion
### If you are using a custom domain
```

> Go to the Network tab and find the failing request (you may need to reload the page, since only requests after opening the developer tools are logged).
> You need to check the response headers for the following two headers:

Reviewer: ```suggestion
1. Open developer tools on your browser.
2. Go to the **Network** tab and find the failing request. You may need to reload the page, as requests are only logged after developer tools have been opened.
3. Check the response headers for the following two headers:
```

- `cf-cache-status`
- `cf-mitigated`

#### If you have a `cf-mitigated` header

Your request was blocked by one of your WAF rules. Inspect your [Security Events](/waf/analytics/security-events/) to identify the cause of the block.

> #### If you don't have a `cf-cache-status` header

Reviewer: ```suggestion
#### If you do not have a `cf-cache-status` header
```

Your request was blocked by [Hotlink Protection](/waf/tools/scrape-shield/hotlink-protection/).

> You need to tweak your Hotlink Protection settings using a [Configuration Rule](/rules/configuration-rules/) or you can disable it completely.

Reviewer: ```suggestion

Edit your Hotlink Protection settings using a [Configuration Rule](/rules/configuration-rules/), or disable it completely.
```

### If you are using the S3 API

> Your request is very likely incorrectly signed.
> You may see a better error message by trying the request over curl.

Reviewer: ```suggestion
Your request may be incorrectly signed. You may obtain a better error message by trying the request over curl.
```

> You can find working S3 signing examples on the [Examples](/r2/examples/aws/) page.

Reviewer: ```suggestion
Refer to the working S3 signing examples on the [Examples](/r2/examples/aws/) page.
```

> ### If it's actually CORS

Reviewer: ```suggestion
### If it is actually CORS
```

Here are some common issues with CORS configurations:

- `ExposeHeaders` is missing headers like `ETag`
- `AllowedHeaders` is missing headers like `Authorization` or `Content-Type`
- `AllowedMethods` is missing methods like `POST`/`PUT`

> ## HTTP 5XX Errors and Capacity limitations of Cloudflare R2

Reviewer: ```suggestion
## HTTP 5XX Errors and capacity limitations of Cloudflare R2
```

> When you encounter an HTTP 5XX error, it's usually a sign that your Cloudflare R2 bucket has been overwhelmed by too many concurrent requests.
> These errors can trigger bucket-wide read and write locks, affecting the performance of all ongoing operations.
> To avoid these disruptions, it's important to implement strategies for managing request volume.

Reviewer: ```suggestion
When you encounter an HTTP 5XX error, it is usually a sign that your Cloudflare R2 bucket has been overwhelmed by too many concurrent requests. These errors can trigger bucket-wide read and write locks, affecting the performance of all ongoing operations.

To avoid these disruptions, it is important to implement strategies for managing request volume.
```

Here are some mitigations you can employ:

> ### Monitor Concurrent Requests

Reviewer: ```suggestion
### Monitor concurrent requests
```

Track the number of concurrent requests to your bucket. If a client encounters a 5XX error, ensure that it retries the operation and communicates with other clients. By coordinating, clients can collectively slow down, reducing the request rate and maintaining a more stable flow of successful operations.

> If your users are directly uploading to the bucket (for example, using the S3 or Workers API), you may not be able to monitor or enforce a concurrency limit. In that case, we recommend Bucket Sharding:

Reviewer: ```suggestion
If your users are directly uploading to the bucket (for example, using the S3 or Workers API), you may not be able to monitor or enforce a concurrency limit. In that case, we recommend bucket sharding.
```

> ### Bucket Sharding

Reviewer: ```suggestion
### Bucket sharding
```

> For higher capacity at the cost of added complexity, consider bucket sharding. This approach distributes reads and writes across multiple buckets, reducing the load on any single bucket.  While sharding won't prevent a single hot object from exhausting capacity, it can mitigate the overall impact and improve system resilience.

Reviewer: ```suggestion
For higher capacity at the cost of added complexity, consider bucket sharding. This approach distributes reads and writes across multiple buckets, reducing the load on any single bucket.  While sharding cannot prevent a single hot object from exhausting capacity, it can mitigate the overall impact and improve system resilience.
```
